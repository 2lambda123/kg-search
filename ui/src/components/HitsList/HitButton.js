/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import "./HitButton.css";

export const HitButton = ({ reference, data, component, onClick }) => {
  const handleClick = (event) => {
    onClick(data, event.currentTarget);
  };

  const Component = component;
  return (
    <button role="link" className="kgs-hit-button" onClick={handleClick} data-reference={reference}>
      <Component data={data} />
    </button>
  );
};

export default HitButton;