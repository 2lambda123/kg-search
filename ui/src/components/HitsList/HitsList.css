/*
 * Copyright 2018 - 2021 Swiss Federal Institute of Technology Lausanne (EPFL)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This open source software code was developed in part or in whole in the
 * Human Brain Project, funded from the European Union's Horizon 2020
 * Framework Programme for Research and Innovation under
 * Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 * (Human Brain Project SGA1, SGA2 and SGA3).
 *
 */

.kgs-hits-list {
    position: relative;
    margin: 0 -20px;
}

@media (min-width: 576px) {
    .kgs-hits-list {
        margin: 0;
    }
}

.kgs-hits-list > ul {
    list-style-type: none;
    display: block;
    -ms-flex-flow: unset;
    flex-flow: unset;
    margin: 0;
    padding: 0;
}

.kgs-hits-list > div {
    border-top: 1px solid var(--color-27);
    width: 100%;
    margin-top: 40px;
    padding: 10px 20px;
    padding-bottom: 20px;
    font-size: var(--font-size-small);
    color: var(--color-27);
}

@media (min-width: 576px) {
    .kgs-hits-list > div {
        padding: 10px 0;
    }
}

.kgs-hits-list > ul > li {
    position: relative;
    display: inline-block;
    width: 100%;
    margin: 0;
}

@media (min-width: 576px) {
    .kgs-hits-list > ul > li {
        margin: 10px 0 10px 0;
    }
}

@media (min-width: 1450px) {
    .kgs-hits-list > ul {
        margin: 0 0 25px 0;
        padding: 0;
    }

    .kgs-hits-list > ul > li {
        width: 100%;
        margin: 0 0 30px 0;
    }
}
