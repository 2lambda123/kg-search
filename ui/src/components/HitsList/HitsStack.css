

.kgs-hits-stack[size="2"] {
  margin-bottom: 15px;
}

.kgs-hits-stack[size="3"] {
  margin-bottom: 20px;
}

.kgs-hits-stack[size="4"] {
  margin-bottom: 25px;
}

.kgs-hits-stack[size="5"] {
  margin-bottom: 30px;
}

.kgs-hits-stack > ul {
  list-style-type: none;
  display: block;
  -ms-flex-flow: unset;
  flex-flow: unset;
  margin: 0;
  padding: 0;
}

.kgs-hits-stack > ul > li {
  position: relative;
  display: inline-block;
  width: 100%;
  margin: 0;
}

@media (min-width: 576px) {
  .kgs-hits-stack > ul > li {
      margin: 10px 0 10px 0;
  }
}

@media (min-width: 1450px) {
  .kgs-hits-stack > ul {
      margin: 0 0 25px 0;
      padding: 0;
  }

  .kgs-hits-stack > ul > li {
      width: 100%;
      margin: 0 0 30px 0;
  }
}

.kgs-hits-stack .kgs-hits-stack__panel {
  background-color: var(--background-color-2);
  box-shadow:
    /* The top layer shadow */
    0 1px 1px rgba(0,0,0,0.15);
}

.kgs-hits-stack[size="2"] .kgs-hits-stack__panel {
  
  box-shadow:
    /* The top layer shadow */
    0 1px 1px rgba(0,0,0,0.15),
    /* The second layer */
    0 10px 0 -5px var(--background-color-9),
    /* The second layer shadow */
    0 10px 1px -4px var(--box-shadow-5);
}

.kgs-hits-stack[size="3"] .kgs-hits-stack__panel {
  box-shadow:
    /* The top layer shadow */
    0 1px 1px rgba(0,0,0,0.15),
    /* The second layer */
    0 10px 0 -5px var(--background-color-9),
    /* The second layer shadow */
    0 10px 1px -4px var(--box-shadow-5),
     /* The third layer */
    0 20px 0 -10px var(--background-color-9),
    /* The third layer shadow */
    0 20px 1px -9px var(--box-shadow-5);
}

.kgs-hits-stack[size="4"] .kgs-hits-stack__panel {
  box-shadow:
    /* The top layer shadow */
    0 1px 1px rgba(0,0,0,0.15),
    /* The second layer */
    0 10px 0 -5px var(--background-color-9),
    /* The second layer shadow */
    0 10px 1px -4px var(--box-shadow-5),
     /* The third layer */
    0 20px 0 -10px var(--background-color-9),
    /* The third layer shadow */
    0 20px 1px -9px var(--box-shadow-5),
    /* The forth layer */
    0 30px 0 -15px var(--background-color-9),
    /* The forth layer shadow */
    0 30px 1px -14px var(--box-shadow-5);
}

.kgs-hits-stack[size="5"] .kgs-hits-stack__panel {
  box-shadow:
  /* The top layer shadow */
  0 1px 1px rgba(0,0,0,0.15),
  /* The second layer */
  0 10px 0 -5px var(--background-color-9),
  /* The second layer shadow */
  0 10px 1px -4px var(--box-shadow-5),
   /* The third layer */
  0 20px 0 -10px var(--background-color-9),
  /* The third layer shadow */
  0 20px 1px -9px var(--box-shadow-5),
  /* The forth layer */
  0 30px 0 -15px var(--background-color-9),
  /* The forth layer shadow */
  0 30px 1px -14px var(--box-shadow-5),
  /* The fifth layer */
  0 40px 0 -20px var(--background-color-9),
  /* The fifth layer shadow */
  0 40px 1px -19px var(--box-shadow-5);
}

@-ms-keyframes chevron-stack__expand-icon {
  0% {
    transform: translateY(0)
  }
  33% {
    transform: translateY(2px)
  }
  66% {
    transform: translateY(-2px)
  }
  100% {
    transform: translateY(0)
  }
}

@-webkit-keyframes chevron-stack__expand-icon {
  0% {
    transform: translateY(0)
  }
  33% {
    transform: translateY(2px)
  }
  66% {
    transform: translateY(-2px)
  }
  100% {
    transform: translateY(0)
  }
}

@keyframes chevron-stack__expand-icon {
  0% {
      transform: translateY(0)
  }
  33% {
    transform: translateY(2px)
  }
  66% {
    transform: translateY(-2px)
  }
  100% {
    transform: translateY(0)
  }
}

.kgs-hits-stack button.kgs-hits-stack__expand-button {
  position: absolute;
  border: 0;
  padding: 1px 8px;
  color: var(--color-19);
  background-color: transparent;
  transition: color .35s ease-in-out;
  width: 100%;
}

.kgs-hits-stack[size="2"] button.kgs-hits-stack__expand-button {
  padding-top: 5px;
}

.kgs-hits-stack[size="3"] button.kgs-hits-stack__expand-button {
  padding-top: 10px;
  
}

.kgs-hits-stack[size="4"] button.kgs-hits-stack__expand-button {
  padding-top: 15px;
}

.kgs-hits-stack[size="5"] button.kgs-hits-stack__expand-button {
  padding-top: 20px;
}

.kgs-hits-stack button.kgs-hits-stack__expand-button:hover {
  color: var(--color-20);
}

.kgs-hits-stack button.kgs-hits-stack__expand-button > svg {
  transition: all .35s ease-in-out;
}

.kgs-hits-stack button.kgs-hits-stack__expand-button:hover > svg {
  color: var(--color-20);
  box-shadow: 1px 1px 2px var(--box-shadow-5);
  -moz-animation: chevron-stack__expand-icon 1s infinite ease-in-out;
  -webkit-animation: chevron-stack__expand-icon 1s infinite ease-in-out;
  animation: chevron-stack__expand-icon 1s infinite ease-in-out;
}