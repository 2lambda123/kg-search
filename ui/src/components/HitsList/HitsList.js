/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import PropTypes from "prop-types";
import "./HitsList.css";
import { HitButton } from "./HitButton";
import { HitsStack } from "./HitsStack";


const HitsListItem = ({item, itemComponent, getKey, onClick }) => {
  if (Array.isArray(item)) {
    const key = getKey(item[0]);
    return (
      <HitsStack reference={key} items={item} itemComponent={itemComponent} getKey={getKey} onClick={onClick} />
    );
  }
  const key = getKey(item);
  return (
    <HitButton reference={key} data={item} component={itemComponent} onClick={onClick} />
  );
};

export const HitsList = ({ className, title, items, itemComponent, getKey, onClick }) => {
  if (!Array.isArray(items) || items.length === 0 || typeof getKey !== "function") {
    return null;
  }
  const classNames = ["kgs-hits-list", className].join(" ");
  return (
    <div className={classNames}>
      {title && (
        <div>{title}</div>
      )}
      <ul>
        {items.map(item => {
          const key = getKey(Array.isArray(item)?items[0]:item);
          return (
            <li key={key}>
              <HitsListItem item={item} itemComponent={itemComponent} getKey={getKey} onClick={onClick} />
            </li>
          );
        })}
      </ul>
    </div>
  );
};

HitsList.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string,
  items: PropTypes.arrayOf(PropTypes.any),
  itemComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object
  ]).isRequired,
  getKey: PropTypes.func.isRequired,
  onClick: PropTypes.func
};

export default HitsList;