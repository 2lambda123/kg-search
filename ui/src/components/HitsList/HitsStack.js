/*
 * Copyright 2018 - 2021 Swiss Federal Institute of Technology Lausanne (EPFL)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This open source software code was developed in part or in whole in the
 * Human Brain Project, funded from the European Union's Horizon 2020
 * Framework Programme for Research and Innovation under
 * Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 * (Human Brain Project SGA1, SGA2 and SGA3).
 *
 */

import React, { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "./HitsStack.css";
import { HitButton } from "./HitButton";

export const HitsStack = ({ items, itemComponent, getKey, onClick }) => {

  const [expanded, setExpanded] = useState(false);

  if (!Array.isArray(items) || !items.length) {
    return null;
  }

  const handleExpandToggle = () => setExpanded(true);

  const item = items[0];
  const key = getKey(item);

  if (expanded) {
    return (
      <div className="kgs-hits-stack">
        <ul>
          {items.map(item => {
            const key = getKey(item);
            return (
              <li key={key}>
                <HitButton reference={key} data={item} component={itemComponent} onClick={onClick} />
              </li>
            );
          })}
        </ul>
      </div>
    );
  }

  return (
    <div className="kgs-hits-stack" size={items.length>5?5:items.length}>
      <div className="kgs-hits-stack__panel">
        <HitButton reference={key} data={item} component={itemComponent} onClick={onClick} />
      </div>
      <button className="kgs-hits-stack__expand-button" onClick={handleExpandToggle} title={items.length>2?`show the ${items.length-1} other versions `:"show the other version"}>
        <FontAwesomeIcon icon="angle-double-down" />
      </button>
    </div>
  );
};

export default HitsStack;