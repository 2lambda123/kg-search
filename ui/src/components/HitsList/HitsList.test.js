/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import uniqueId from "lodash/uniqueId";
import renderer from "react-test-renderer";
import Enzyme, { mount, shallow, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import HitsList from "./HitsList";

Enzyme.configure({ adapter: new Adapter() });

test('HitsList component renders initially', () => {
    const component = renderer.create(
        <HitsList className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => (<div></div>)} />
    );
  
    expect(component.toJSON()).toMatchSnapshot();
});
  
test('HitsList test className"', () => {
    const component = shallow(
        <HitsList className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => (<div></div>)} />
    );
    expect(component.hasClass("className"));
});

test('HitsList test title"', () => {
    const component = shallow(
        <HitsList className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => (<div></div>)} />
    );
    expect(component.find("div div").text()).toBe("a title");
});

test('HitsList test number of items', () => {
    const component = render(
        <HitsList className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={() => {}} itemComponent={() => (<div></div>)} />
    );
    expect(component.find("li").length).toBe(2);
});

test('Toggle test  button click', () => {
    const fn = jest.fn();
    const component = mount(
        <HitsList className="className" title="a title" items={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} getKey={() => uniqueId()} onClick={fn} itemComponent={() => (<div></div>)} />
    );
    component.find('button').at(1).simulate('click');
    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0].value).toBe("another value");
});