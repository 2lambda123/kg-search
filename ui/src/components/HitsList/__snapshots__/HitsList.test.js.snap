// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HitsList component renders initially 1`] = `
<div
  className="kgs-hits-list className"
>
  <div>
    a title
  </div>
  <ul>
    <li>
      <button
        className="kgs-hit-button"
        data-reference="3"
        onClick={[Function]}
        role="link"
      >
        <div />
      </button>
    </li>
    <li>
      <button
        className="kgs-hit-button"
        data-reference="4"
        onClick={[Function]}
        role="link"
      >
        <div />
      </button>
    </li>
  </ul>
</div>
`;
